Source/Metal/CommandBufferMTL.mm

```diff
@@ -0,0 +1,119 @@
+void CommandBufferMTL::SetConstants(uint32_t pushConstantIndex, const void* data, uint32_t size) {}
+void CommandBufferMTL::SetDescriptorPool(const DescriptorPool& descriptorPool) {}
+void CommandBufferMTL::Barrier(const BarrierGroupDesc& barrierGroupDesc) {}
+void CommandBufferMTL::BeginRendering(const AttachmentsDesc& attachmentsDesc) {
+    m_encoder = [m_Handle renderCommandEncoderWithDescriptor: NULL];
+    
+}
+void CommandBufferMTL::EndRendering() {}
+void CommandBufferMTL::SetViewports(const Viewport* viewports, uint32_t viewportNum) {}
+void CommandBufferMTL::SetScissors(const Rect* rects, uint32_t rectNum) {}
+void CommandBufferMTL::SetDepthBounds(float boundsMin, float boundsMax) {}
+void CommandBufferMTL::SetStencilReference(uint8_t frontRef, uint8_t backRef) {}
+void CommandBufferMTL::SetSamplePositions(const SamplePosition* positions, Sample_t positionNum, Sample_t sampleNum) {}
+void CommandBufferMTL::SetBlendConstants(const Color32f& color) {}
+void CommandBufferMTL::SetShadingRate(const ShadingRateDesc& shadingRateDesc) {}
+void CommandBufferMTL::ClearAttachments(const ClearDesc* clearDescs, uint32_t clearDescNum, const Rect* rects, uint32_t rectNum) {}
+void CommandBufferMTL::SetIndexBuffer(const Buffer& buffer, uint64_t offset, IndexType indexType) {
+    m_CurrentIndexCmd.m_Buffer = &(BufferMTL&)buffer;
+    switch(indexType) {
+        case IndexType::UINT16:
+            m_CurrentIndexCmd.m_Type = MTLIndexType::MTLIndexTypeUInt16;
+            break;
+        default:
+        case IndexType::UINT32:
+            m_CurrentIndexCmd.m_Type = MTLIndexType::MTLIndexTypeUInt32;
+            break;
+    }
+    m_CurrentIndexCmd.m_Offset = offset;
+}
+void CommandBufferMTL::SetVertexBuffers(uint32_t baseSlot, uint32_t bufferNum, const Buffer* const* buffers, const uint64_t* offsets) {
+}
+void CommandBufferMTL::Draw(const DrawDesc& drawDesc) {
+    
+    [m_encoder drawPrimitives: m_CurrentPipeline->m_graphics.m_primitiveType
+            vertexStart:drawDesc.baseVertex
+            vertexCount:drawDesc.vertexNum
+            instanceCount:drawDesc.instanceNum
+               baseInstance:0];
+}
+void CommandBufferMTL::DrawIndexed(const DrawIndexedDesc& drawIndexedDesc) {
+    id<MTLBuffer> indexBuffer = m_CurrentIndexCmd.m_Buffer->GetHandle();
+    [m_encoder drawIndexedPrimitives: m_CurrentPipeline->m_graphics.m_primitiveType
+                              indexCount:drawIndexedDesc.indexNum
+                               indexType: m_CurrentIndexCmd.m_Type
+                             indexBuffer: indexBuffer
+                       indexBufferOffset: m_CurrentIndexCmd.m_Offset];
+}
+void CommandBufferMTL::DrawIndirect(const Buffer& buffer, uint64_t offset, uint32_t drawNum, uint32_t stride, const Buffer* countBuffer, uint64_t countBufferOffset) {}
+void CommandBufferMTL::DrawIndexedIndirect(const Buffer& buffer, uint64_t offset, uint32_t drawNum, uint32_t stride, const Buffer* countBuffer, uint64_t countBufferOffset) {}
```

Comment:
looks like there isn't a count buffer implement provided by the encoder. https://developer.apple.com/documentation/metal/mtlrendercommandencoder/render_pass_drawing_commands?language=objc

Reply:
Ping me when you're done with the base implementation, I'll try to get samples to work.

Reply:
Also, argument buffers can be used to emulate root constants/push constants. But yes, you can also use it for count buffers

Reply:
@vertver I think i have the core pieces in place still not sure about the overall implementation.